lib/puppet/pops/types/tree_iterators.rb

```diff
@@ -0,0 +1,246 @@
+module Puppet::Pops::Types
+module Iterable
+
+class TreeIterator
+  include Iterable
+
+  DEFAULT_CONTAINERS = Puppet::Pops::Types::TypeFactory.variant(
+    Puppet::Pops::Types::PArrayType::DEFAULT,
+    Puppet::Pops::Types::PHashType::DEFAULT,
+    Puppet::Pops::Types::PObjectType::DEFAULT
+    )
+
+  # Creates a TreeIterator that by default treats all Array, Hash and Object instances as
+  # containers - the 'containers' option can be set to a type that denotes which types of values
+  # should be treated as containers - a `Variant[Array, Hash]` would for instance not treat
+  # Object values as containers, whereas just `Object` would only treat objects as containers.
+  #
+  # Unrecognized options are silently ignored
+  #
+  # @param [Hash] options the options
+  # @option options [PType] :container_type ('Variant[Hash, Array, Object]') The type(s) that should be treated as containers. The
+  #   given type(s) must be assignable to the default container_type.
+  # @option options [Boolean] :include_root ('true') If the root container itself should be included in the iteration (requires
+  #   `include_containers` to also be `true` to take effect).
+  # @option options [Boolean] :include_containers ('true') If containers should be included in the iteration
+  # @option options [Boolean] :include_values ('true') If non containers (values) should be included in the iteration
+  # @option options [Boolean] :include_refs ('false') If (non containment) referenced values in Objects should be included 
+  #
+  def initialize(enum, options={})
+    @root = enum
+    @element_t = nil
+    @value_stack = [enum]
+    @indexer_stack = []
+    @current_path = []
+    @recursed = false
+    @containers_t = options['container_type'] || DEFAULT_CONTAINERS
+    unless DEFAULT_CONTAINERS.assignable?(@containers_t)
+      raise ArgumentError, _("Only Array, Hash, and Object types can be used as container types. Got %{type}") % {type: @containers_t}
+    end
+    @with_root = options['include_root'].nil? ? true : !!options['include_root']
+    @with_containers = options['include_containers'].nil? ? true : !!options['include_containers']
+    @with_values = options['include_values'].nil? ? true : !!options['include_values']
+    @with_root = @with_containers && (options['include_root'].nil? ? true : !!options['include_root'])
```

Comment:
Lots of repeated behavior here with an extra lookup for each key. Could perhaps be replaced by a more efficient method such as:
```ruby
def extract_option(options, key, default)
  v = options[key]
  v.nil? ? default : !!v
end
```